compaction/compaction.cc

```diff
@@ -864,25 +878,93 @@ void compacted_fragments_writer::maybe_abort_compaction() {
     }
 }
 
-void compacted_fragments_writer::consume_new_partition(const dht::decorated_key& dk) {
+void compacted_fragments_writer::stop_current_writer() {
+    // stop sstable writer being currently used.
+    _stop_compaction_writer(&*_compaction_writer);
+    _compaction_writer = std::nullopt;
+}
+
+bool compacted_fragments_writer::can_split_large_partition() const {
+    return _c._can_split_large_partition;
+}
+
+void compacted_fragments_writer::split_large_partition() {
+    // Closes the active range tombstone if needed, before emitting partition end.
+    if (_current_partition.current_emitted_tombstone) {
+        auto rtc = range_tombstone_change(position_in_partition::after_key(_current_partition.last_pos), tombstone{});
+        _c.log_debug("Closing active tombstone {} with {} for partition {}", _current_partition.current_emitted_tombstone, rtc, *_current_partition.dk);
+        _compaction_writer->writer.consume(std::move(rtc));
+    }
+    _c.log_debug("Splitting large partition {} in order to respect SSTable size limit of {}", *_current_partition.dk, pretty_printed_data_size(_c._max_sstable_size));
+    // Close partition in current writer, and open it again in a new writer.
+    do_consume_end_of_partition();
+    stop_current_writer();
+    do_consume_new_partition(*_current_partition.dk);
+    // Replicate partition tombstone to every fragment, allowing the SSTable run reader
+    // to open a single fragment during the read.
+    if (_current_partition.tombstone) {
+        consume(_current_partition.tombstone);
+    }
```

Comment:
You have to re-emit the active tombstone here, with the same position that you used for closing (and the proper tombstone).

Reply:
after_key(last_pos) to avoid introducing overlapping at clustering domain right?

Reply:
The upper layer (mutation_compactor or whatever) will still emit the "close marker" for the open-ended range tombstone, before emitting end_of_partition, that we decided to split at the inner layer (compacted fragment writer). Correct? So the splitting is necessary exactly for the reason of keeping the disjoint invariant. This should be well documented or I will forget it the next day :-)

Reply:
Of course, also for correctness.